Issue 644 Support for multiple AGWs/LBs and named backend pools

Justin-W opened the pull request:
- [x] Please check this box and fill the data as below once you have run the unit tests.
      Code coverage should keeps at 100%. 

  ```
  pushd src/bosh_azure_cpi
    ./bin/test-unit
  popd
  ...
  Finished in 5.25 seconds (files took 0.6538 seconds to load)
  1019 examples, 0 failures, 9 pending
  
  Coverage report generated for RSpec to /Users/justinw/go/src/github.com/cloudfoundry/bosh-azure-cpi-release/src/bosh_azure_cpi/coverage. 4247 / 4266 LOC (99.55%) covered.
  ```

### Testing Notes:

- All of the new code added by this PR is fully covered by unit specs.
- I didn't add integration specs for the new features, since it would involve either modifying the existing integration and BATS assets, or else adding a parallel separate set of such assets.
- The 9 Pending specs are new specs that I added, which relate to existing (long before this PR) uncovered lines of code. I didn't have time to implement these 9, but I figured adding these as pending specs would make it easier fo someone else to get the project back up to 100% coverage.
  * I did add some new specs (which ar unrelated to this PR), to cover some of the pre-existing uncovered lines. The 9 remaining Pending specs are the ones left over.

## Rubocop:

  ```
  bosh_azure_cpi % ./bin/rubocop_check
  ~/go/src/github.com/cloudfoundry/bosh-azure-cpi-release/src/bosh_azure_cpi 
  ~/go/src/github.com/cloudfoundry/bosh-azure-cpi-release/src/bosh_azure_cpi
  Running rubocop with Ruby version ruby-2.7.3...
  ...
  187 files inspected, no offenses detected
  ```

### Rubocop Notes:

- I fixed some pre-existing rubocop offenses (mostly related to layout and formatting).
  * I was going to fix more, but was worried that it would affect so many files it would make this PR hard to code review. _If this PR gets accepted fairly soon, I might have time left to submit additional rubocop fixes as a separate PR._

### Changelog

* Fixes #644.
* Modified the `vm_type/application_gateway` config to allow an Array (of Hashes) to be specified (as an alternative to the currently-supported config data types of String and Hash), to enable multiple AGWs (with independent `vm_type/application_gateway/resource_group_name` or `vm_type/application_gateway/backend_pool_name` values).
* Modified the `vm_type/load_balancer` config to support a new `vm_type/load_balancer/backend_pool_name` sub-property, to allow the config to exlpicitly specify the name of the backend address pool to use. (This change re-establishes/maintains symmetry between the `vm_type/load_balancer` and `vm_type/application_gateway` configs.)
* Modified the `vm_type/load_balancer` config to allow an Array (of Hashes) to be specified, to enable multiple load balancers (with independent `vm_type/load_balancer/resource_group_name` or `vm_type/load_balancer/backend_pool_name` values). (This change re-establishes/maintains symmetry between the `vm_type/load_balancer` and `vm_type/application_gateway` configs.)
* Fixed various missing and inaccurate doc comments.
* Fixed various rubocop offenses.
* Fixed the names/contexts of some existing specs with inaccurate names.
* Added missing specs for previously-uncovered code.
* Upgraded to RSpec from v.5.0 to v3.10.0 (and revendored the gems).

klakin-pivotal commented:
If you're not officially done with this PR and submitting it for review, please ignore the remainder of this comment.

I'm _probably_ not qualified to be a _good_ reviewer for this.

But. If you're officially done with this PR, would you squash the >150 commits in this PR into one or more commits that group the changes into more-easily-understandable changesets? Thanks!

Justin-W commented:
> If you're officially done with this PR, would you squash the >150 commits in this PR into one or more commits that group the changes into more-easily-understandable changesets? Thanks!

@klakin-pivotal I am officially done, and ready for the PR to be reviewed. And I'd be happy to squash the commits down _if_ you want.

However, before I squash any further, please note that the current number of commits (169) is _after_ I had already squashed a lot of the intermediate commits, and re-ordered most of the commits into sequential, logical groups. However, some of the files changed (particularly some of the spec files) resulted in some pretty significant changes as a final result of the incremental commits. IMO, squashing some of the remaining incremental commits would make it significantly harder (or impossible) to understand the incremental changes that led to the major changes. And git tools do make it fairly easy to view diffs of arbitrary ranges of commits, but can't show you the incremental commits again after they've been squashed.

I also spent a fair amount of effort on making sure that the individual commits' messages are accurate, to make analysis and review of the PR easier.

So with that said, I'd still be happy to squash the PR's commits down (a little or a lot, as you prefer), _if_ you still want me to. But it is a one-way process that seems to have more drawbacks than benefits (when considering what is easily achievable with git CLI or git clients), so I wanted to double check that you really want me to do so (and by how much) before I do it, since if I were the one reviewing such a PR, I'd prefer more, smaller, simpler commits over fewer, larger, squashed commits. But if you can (re)confirm your preference, I'll do it. 😄

klakin-pivotal commented:
> However, some of the files changed (particularly some of the spec files) resulted in some pretty significant changes as a final result of the incremental commits...

Are _you_ certain that -for example- this group of four commits:

<https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/150a0d25ec937f25bd77b75e9a5a92d1f5543a35>
<https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/3a70ca989f8def47ae655f999f121f7779c8816f>
<https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/f188d0977170d310e20f31c74ddc3c7bd4dac67b>
<https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/874d7e4b81dce9d7729cd93ae2ce44c3c57ef331>

or this group of two commits:
<https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/8cda7590a37fdd2444d905b85cac1ba8198394b3>
<https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/42de3163b911a3f161216998f391f179cb316a7d>

aren't sufficiently related that the commits in each of the two groups should be squashed down to one commit per group?

If -after reviewing the commits I called out-, you do change your mind about not squashing them together, then I'd politely request that you take an hour or two to do some history grooming on this PR... squashing commits together that _are_ logically related, and even reordering the commits so that it's easier to squash together commits that are logically related but separated by other -unrelated- commits. Squashing and -where appropriate- reordering commits will better ensure that the story your remaining commits do tell makes as much sense as is reasonably possible.

Additionally, it's possible that you and I are looking at two different sets of commits. If we are, then please accept my apologies for the noise, and please do direct me to the list of commits that I should be looking at.

Justin-W commented:
> I'd politely request that you take an hour or two to do some history grooming on this PR... squashing commits together that _are_ logically related, and even reordering the commits so that it's easier to squash together commits that are logically related but separated by other -unrelated- commits. Squashing and -where appropriate- reordering commits will better ensure that the story your remaining commits do tell makes as much sense as is reasonably possible.

@klakin-pivotal Done. I think it now presents a more easily understood story.  If you'd like me to squash further, I certainly can, but doing so may affect the ability to easily revert/amend isolated portions of the PR's changes, if any of the reviewers request changes to a particular design/implementation decision. Also, see notes below.

### Notes about the commits in this PR:

* Of the 84 commits, only the 40 commits with `issue-644` in the commit msg are directly related to new code/functionality for #644.
* The other 44 commits are related to existing code, and are mostly cleanup/refactoring/enhancement of code sections that relate to or were subsequently modified by the commits directly relating to #644. Most of those 44 commits are clustered at the beginning (before the first `issue-644` commit).
* Just in case, I temporarily preserved the original (pre-squashed) version of this PR's commit history [here](https://github.com/Justin-W/bosh-azure-cpi-release/commits/pr-651-v1).

beyhan commented on src/bosh_azure_cpi/spec/unit/vm_manager/create/shared_stuff.rb:
```diff
@@ -111,14 +121,42 @@
+      ]
     }
   end
+  # let(:network_interfaces) do
```
Why this commented out code is there?

Justin-W replied:
A number of existing specs have `let(:network_interfaces)` declarations. I was playing around with adding one here in the shared `let`s to consolidate (at least some of) these duplicated code sections, but I didn't complete that change.

I can remove this section, or leave it as a reminder of an opportunity for future refactoring, as you prefer.

beyhan replied:
I you would like to complete this please do that. Otherwise I would suggest to remove the commented out code.

Justin-W replied:
Done.

beyhan commented on src/bosh_azure_cpi/spec/unit/vm_manager/create/shared_stuff.rb:
```diff
@@ -53,7 +53,17 @@
       'storage_account_name' => 'dfe03ad623f34d42999e93ca',
       'caching' => 'ReadWrite',
       'load_balancer' => 'fake-lb-name',
+      # 'load_balancer' => {
```
Why this commented out code is there?

Justin-W replied:
Because I wasn't sure whether it was better to change the shared `let`s to use the new Hash or Array<Hash> format for the `load_balancer` and `application_gateway` config in the `:vm_properties`, or leave them using the older, original String-typed values.

All existing specs are passing either way, but I left the original default values as-is, and added the equivalent new-format values as commented-out alternatives.

What would you like done here?

beyhan replied:
I would suggest to remove the commented out code.

Justin-W replied:
Done.

beyhan commented on src/bosh_azure_cpi/spec/unit/utils/bosh_agent_util_spec.rb:
```diff
@@ -33,77 +32,102 @@
-          dns: { nameserver: dns }
-      }
-    end
+  # TODO: coverage: need to implement the following specs to increase coverage
```
Are the TODOs still required?

Justin-W replied:
Yes, these TODOs make explicit some areas of un-covered code that are already in `master`, and are technically unrelated to this PR (other than the fact that they bing coverage down for this and all other PRs until they're fixed and coverage returns to 100%, which is stated as a requirement in the template for all PRs).

Justin-W replied:
FYI: More info on these TODOs is in #652.

beyhan commented on src/bosh_azure_cpi/lib/cloud/azure/models/vm_cloud_props.rb:
```diff
@@ -99,21 +102,69 @@ def _default_fault_domain_count(global_azure_config)
+
+      return nil unless load_balancer_config
+
+      cloud_error("Property '#{LOAD_BALANCER_KEY}' must be a String, Hash, or Array.") unless load_balancer_config.is_a?(String) || load_balancer_config.is_a?(Hash) || load_balancer_config.is_a?(Array)
```
Maybe we can have a generic function which parses the property. It can return a structure back like [{names => names, group_name => group_names, pool_name => pool_names}]. This output can be used to instantiate the corresponding  `Config` here. We could use this for the load balancer and application gateway. Just a suggestion.

Justin-W replied:
The 2 methods' implementations are currently slightly distinct (e.g. in the handling of `nil` vs defaulted values for the `resource_group_name`), and I'm not sure how much benefit there would currently be in consolidating only these 2 methods. If additional similar methods are needed in future, I'd agree that some de-duping via helper methods would probably be better.

beyhan replied:
fine with me.

beyhan reviewed (commented):
I also couldn't get into all the details here. I have some small suggestion otherwise the feature implementation looks good. One additional question. Did you consider to extend the integration tests or the unit tests provide good coverage here for the introduced changes?

Justin-W commented on src/bosh_azure_cpi/lib/cloud/azure/cloud.rb:
```diff
@@ -22,7 +23,7 @@ class Cloud < Bosh::Cloud
     #
     # @param [Hash] options cloud options
-    def initialize(options, api_version)
+    def initialize(options, api_version = 1)
```
The default value in the signature only applies when the optional arg is omitted by the caller. If the caller explicitly specifies the arg with a value of `nil`, the arg default doesn't override the `nil`, but the ` || 1` on line 29 will.

And without the arg default on linme 26, the arg technically isn't optional, it is required. Meaning that it is technically an error for any caller to not pass that arg, meaning line 29 is covering up the caller's misuse of the method. 

So both "defaults" are warranted.

beyhan replied:
good point with passing the a `nil` value.

Justin-W commented:
> One additional question. Did you consider to extend the integration tests or the unit tests provide good coverage here for the introduced changes?

Yes and yes.

The new and updated unit specs provide much more thorough coverage (of both the new and old formats) of the various config variations than I saw for any other properties' unit specs. E.g. Single vs multiple LB/AGW, single vs multiple backend pools, unspecified vs explicit pool names, and specs for each supported "data format" (including already-supported formats which didn't have any unit nor integration spec coverage before.

And yes, I considered adding integration specs, as well. I had started some preliminary work in [87b33ed](https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/87b33ed2c892ced6d775566d969b99f6e82bf9bd), but reverted it in [e5e3b1c](https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/e5e3b1c86292a01a5435fa475d32cab5b0c6a3c1), because (as noted in the Testing Notes [here](https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651#issue-800226173):

> I didn't add integration specs for the new features, since it would involve either modifying the existing integration and BATS assets, or else adding a parallel separate set of such assets.

And since there is now (due to this PR's enhancements) excellent unit spec coverage of the LB and AGW config variants, and since I also wasn't sure if anyone other than my team would even be using the new feature, and since none of the previous LB variant format enhancements had updated the integration specs (and some hadn't even updated the unit specs nor the CPI docs), it seemed like it would probably be unnecessary and/or overkill to implement an entire parallel set of int specs (especially since the int suite costs real money per run).

Plus, the initially-broken dev env setup instructions, LB-related refactoring, retroactive specs, and other repo cleanup work I ended up having to do for this PR already took a lot more time than I had planned on spending on this entire PR. 😄

RobertBoyd commented:
Is there any progress on the reviews against this PR. Would really like to take advantage of this new functionality...

Thanks!

klakin-pivotal commented:
I have some things that I want to say about the PR, but we've been quite busy with a few very urgent fires. Will hopefully get around to making some comments next week.

beyhan commented on src/bosh_azure_cpi/spec/unit/azure_client/list_network_interfaces_by_keyword_spec.rb:
```diff
@@ -26,6 +26,28 @@
   let(:valid_access_token) { 'valid-access-token' }
   let(:expires_on) { (Time.new + 1800).to_i.to_s }
 
+  describe '#get_resources_by_url' do
```
Why we have the describe `#get_resources_by_url`  with tests which are empty?

Justin-W replied:
This code is related to #652 (areas of un-covered code which are already in master), and is technically unrelated to either #644 or this PR (other than the fact that they bring coverage down for this and all other PRs until #652 is fixed and coverage returns to 100%, which is stated as a requirement in the template for all PRs).

I had started an attempt to return coverage to 100% within this PR, but found that doing so would take too much time. So I left 9 Pending/unimplemented specs to be addressed later separately (see #652 for details).

beyhan replied:
I will suggest to remove the empty test cases because we want to change the requirement to have 100% code coverage.

Justin-W replied:
removed as requested

beyhan commented on src/bosh_azure_cpi/spec/unit/utils/bosh_agent_util_spec.rb:
```diff
@@ -33,77 +32,102 @@
-    end
+  # TODO: coverage: need to implement the following specs to increase coverage
+  #   The uncovered code seems to be related to commit #c38b3026f9512491ffcb70be7088b4d3e319fdc0 (for issue #491).
+  describe '#encoded_user_data' do # rubocop:disable RSpec/RepeatedExampleGroupBody
```
Why is this test case empty?

Justin-W replied:
This code is related to #652 (areas of un-covered code which are already in master), and is technically unrelated to either #644 or this PR (other than the fact that they bring coverage down for this and all other PRs until #652 is fixed and coverage returns to 100%, which is stated as a requirement in the template for all PRs).

I had started an attempt to return coverage to 100% within this PR, but found that doing so would take too much time. So I left 9 Pending/unimplemented specs to be addressed later separately (see #652 for details).

beyhan replied:
I will suggest to remove the empty test cases because we want to change the requirement to have 100% code coverage.

Justin-W replied:
removed as requested

beyhan commented on src/bosh_azure_cpi/spec/unit/utils/bosh_agent_util_spec.rb:
```diff
@@ -33,77 +32,102 @@
-      )
+  # TODO: coverage: need to implement the following specs to increase coverage
+  #   The uncovered code seems to be related to commit #c38b3026f9512491ffcb70be7088b4d3e319fdc0 (for issue #491).
+  describe '#encode_user_data' do # rubocop:disable RSpec/RepeatedExampleGroupBody
```
Same: why is this test case empty?

Justin-W replied:
This code is related to #652 (areas of un-covered code which are already in master), and is technically unrelated to either #644 or this PR (other than the fact that they bring coverage down for this and all other PRs until #652 is fixed and coverage returns to 100%, which is stated as a requirement in the template for all PRs).

I had started an attempt to return coverage to 100% within this PR, but found that doing so would take too much time. So I left 9 Pending/unimplemented specs to be addressed later separately (see #652 for details).

beyhan replied:
I will suggest to remove the empty test cases because we want to change the requirement to have 100% code coverage.

Justin-W replied:
removed as requested

beyhan commented on src/bosh_azure_cpi/spec/unit/utils/bosh_agent_util_spec.rb:
```diff
@@ -33,77 +32,102 @@
-    end
+  # TODO: coverage: need to implement the following specs to increase coverage
+  #   The uncovered code seems to be related to commit #c38b3026f9512491ffcb70be7088b4d3e319fdc0 (for issue #491).
+  describe '#meta_data_obj' do # rubocop:disable RSpec/RepeatedExampleGroupBody
```
Same: why is this test case empty?

Justin-W replied:
This code is related to #652 (areas of un-covered code which are already in master), and is technically unrelated to either #644 or this PR (other than the fact that they bring coverage down for this and all other PRs until #652 is fixed and coverage returns to 100%, which is stated as a requirement in the template for all PRs).

I had started an attempt to return coverage to 100% within this PR, but found that doing so would take too much time. So I left 9 Pending/unimplemented specs to be addressed later separately (see #652 for details).

RobertBoyd replied:
Any update on this PR? Would really like to take advantage of this feature in Azure.

Thanks!

beyhan replied:
I will suggest to remove the empty test cases because we want to change the requirement to have 100% code coverage.

Justin-W replied:
removed as requested

beyhan reviewed (commented):
@Justin-W Thank you for the fast reaction on my comments. I added additional comments about empty test cases. Sorry, for missing those during the first review. I also adde @klakin-pivotal as reviewer because of the comment here https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651#issuecomment-997107004

beyhan reviewed (changes requested):
I as commented we should remove the empty test cases which aren't related to this pr. Otherwise it looks good to me.

Justin-W commented:
> I as commented we should remove the empty test cases which aren't related to this pr. Otherwise it looks good to me.

@beyhan I have removed all of the Pending specs, as you requested above. And I believe that all of the previous PR comments are already resolved.

RobertBoyd commented:
> If you're not officially done with this PR and submitting it for review, please ignore the remainder of this comment.
> 
> I'm _probably_ not qualified to be a _good_ reviewer for this.
> 
> But. If you're officially done with this PR, would you squash the >150 commits in this PR into one or more commits that group the changes into more-easily-understandable changesets? Thanks!


Hi everyone, I'm wondering whether this PR can be approved or there are additional issues blocking merge?

beyhan commented:
Hi @RobertBoyd,
@klakin-pivotal wanted to provide additional feedback to the pr. There is nothing else blocking the merge and it is approved by me and @ramonskie.

klakin-pivotal commented:
I've carved out a three hour block at the end of my day today to do PR review... so I'll give this a look and add comments.

For reference, I'm in US Pacific time.

klakin-pivotal commented:
(Apologies for the huge delay, the two weeks leading up to the Christmas break were _insanely_ busy.)

RobertBoyd commented:
> I've carved out a three hour block at the end of my day today to do PR review... so I'll give this a look and add comments.
> 
> For reference, I'm in US Pacific time.

Thanks for responding @klakin-pivotal. We're trying to wrap up this thread of work and appreciate the review - @Justin-W and myself are also in Pacific Time.

klakin-pivotal commented:
In terms of total commit count, if merged, this PR would make up ~9% of the commits in this repo. I've looked over the PR, and I don't understand _why_ there are so many commits.

I would _never_ reject a patch series just because the best way to split it up is into a whole bunch of commits... if you need 100 or 1000 commits to break some work up into the most logical series of changes, then that's _definitely_ what needs done.

However, there are two big things about this patch series that bother me, and several small things.

### The big things:

1) It contains many, many, many commits that have a brief explanation of _what_ was done but absolutely no explanation of why. Not _every_ commit needs to explain the why... but nearly zero of your commits explain why what was done was done.
2) There are a bunch of commits that appear to be pretty spurious. For example: several commits in the patch series that (according to the commit message) _just_ revert previous commits. Is it _important_ to someone who will look at this commit history twelve months down the road to know about things like that?

### The small things:

1) These "MAYBE?" commits? Is that commit message the correct thing to be telling a future someone who's trying to understand the _why_ of this patch series? Do they need to understand how uncertain you were of that particular commit? If you think they _do_, I would consider that a signal that commits like that should be squashed into one or more other commits that (when combined) made you sufficiently certain of your changes to submit them as a PR.
2) What's up with the Rubocop rule changes? Why were _any_ Rubocop rule changes required?
3) Why does this patch series contain a change to the version of rspec? Was that a _required_ part of adding this feature? If it wasn't, then that change should not be part of this patch series.
4) Some of this patch series has parts that _just_ add or remove a blank line. These changes should not be in this patch series.
5) What's up with the return type annotations that are added to _some_ functions? For an example, consider the additional annotations that begin here: <https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/files#diff-6eaadd97bb87d9f3b51e3214a0aa347de08a67afae0923b61b84cad6073b50b4R1955> Much of the rest of the file doesn't appear to have those... why were they added to otherwise untouched code? If you wanted to add type annotations to all of the functions in the CPI, that would be _lovely_... but that sort of change should be in a separate PR, not in this one.
6) Does this entirely-commented-out block actually do anything? <https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/files#diff-1a1fbe20391f7dce08027421a87d3efb885328926612e3823ccf57a05259b7f0R172> If it does, I _really_ wanna know... Ruby is magical sometimes, and I'm always excited/horrified to learn of yet another way it can appear to do one thing, but actually do another. If it doesn't actually do something, you should remove it and also do a careful sweep to ensure that you haven't put _other_ commented-out dead code into this patch series.
7) Looking at the individual commits, I would have an _extremely_ difficult time reviewing this PR. In fact, I had to have GitHub squash them down into a single diff so that I could understand the changes. That is a _very strong sign_ that anyone twelve months in the future who's trying to make sense of this part of the Git history will have a _very bad time_.

An ideal commit should contain changes that are logically grouped together, with an explanation of what was done and why. This helps future folks who don't have any context on what was changed in the code and why (which -BTW- this future person _very often_ ends up being _you_) understand the what and why of the code.

Please do squash this PR down to just a few commits that someone who has nearly-zero context on _what_ was done in this PR and _why_ it was done can read and understand. Remove the unnecessary whitespace changes, either help me to understand why you made RuboCop rules changes, or just undo those changes, and address the other stuff I brought up.

_For now, I strongly NAK merging this PR as-is._

As an aside: if you're like me, and you make a ton of "checkpoint" commits while working on something big and/or unfamiliar, do feel free to keep those in a local branch for as long as you like. I -personally- find that keeping those commits around for quite a while _does_ help _me_ to find and fix errors in the code that I wrote... but I also recognize that the very many commits are only helpful to me _because_ I remember much of what I was thinking about when I authored those commits. I _also_ recognize that that mess of commits are _nearly completely_ useless to folks who lack that context. Other folks need the changes bundled up in larger chunks, with notes attached in the commit messages that explain all of the context that I had in my head while I was making those changes.

rkoster commented:
@klakin-pivotal I'm unfamiliar with NAK acronym. Does `For now, I strongly NAK merging this PR as-is.` mean you are okeay with this PR being merged as is?

bgandon commented:
I definitely agree with @klakin-pivotal and will position myself as against merging such a way-too-long list of commits, even though the added feature is fine. As Kenneth said “_commit should contain changes that are logically grouped together_”. 30 years I've been writing software, I could not tell better.

Please follow carefully Kenneth advices. It's rare that a very experienced software engineer take so much effort and time in detailing a Git history problem and how to get over it. So, take the opportunity for learning @Justin-W, it's a great chance you have here!

Finally, I'll just give you these concrete hints: use `git rebase -i` with `fixup` or `squash` strategies in order to rewrite this history of commit properly, then `git push --force-with-lease` appropriately.
Thank you!

bgandon reviewed (changes requested):
I definitely agree with @klakin-pivotal and will position myself as against merging such a way-too-long list of commits, even though the added feature is fine. As Kenneth said “_commit should contain changes that are logically grouped together_”. 30 years I've been writing software, I could not tell better.

Please follow carefully Kenneth advices. It's rare that a very experienced software engineer take so much effort and time in detailing a Git history problem and how to get over it. So, take the opportunity for learning @Justin-W, it's a great chance you have here!

Finally, I'll just give you these concrete hints: use `git rebase -i` with `fixup` or `squash` strategies in order to rewrite this history of commit properly, then `git push --force-with-lease` appropriately.
Thank you!

klakin-pivotal commented:
> @klakin-pivotal I'm unfamiliar with NAK acronym. Does `For now, I strongly NAK merging this PR as-is.` mean you are okeay with this PR being merged as is?

It is the opposite of that. I am _not_ okay with the PR being merged. It's the opposite of ACK, and in wide use in the circles that I hang out in (and _many_ others that I do not). Apologies for the confusion.

Justin-W commented:
Thanks for the detailed feedback @klakin-pivotal. I believe I have now addressed your concerns (details below).

> In terms of total commit count, if merged, this PR would make up ~9% of the commits in this repo. I've looked over the PR, and I don't understand _why_ there are so many commits.

I have squashed most of the original commits down to 6 commits, and removed all other unrelated and "housekeeping" commits (see below for details) which seemed to be "non-essential" to this PR.

> I would _never_ reject a patch series just because the best way to split it up is into a whole bunch of commits... if you need 100 or 1000 commits to break some work up into the most logical series of changes, then that's _definitely_ what needs done.
> 
> However, there are two big things about this patch series that bother me, and several small things.
> 
> ### The big things:
> 1. It contains many, many, many commits that have a brief explanation of _what_ was done but absolutely no explanation of why. Not _every_ commit needs to explain the why... but nearly zero of your commits explain why what was done was done.

All 6 remaining commits now have messages explaining the "why" behind the "what".

> 2. There are a bunch of commits that appear to be pretty spurious. For example: several commits in the patch series that (according to the commit message) _just_ revert previous commits. Is it _important_ to someone who will look at this commit history twelve months down the road to know about things like that?

The "MAYBE?" commit (and the following commit which reverted it) have been removed (refer to the explanation below for details, if interested).

> ### The small things:
> 1. These "MAYBE?" commits? Is that commit message the correct thing to be telling a future someone who's trying to understand the _why_ of this patch series? Do they need to understand how uncertain you were of that particular commit? If you think they _do_, I would consider that a signal that commits like that should be squashed into one or more other commits that (when combined) made you sufficiently certain of your changes to submit them as a PR.

TL;DR: I have removed that code from this PR.

Background:

This particular commit was related to the Integration test suite, and my uncertainty of whether or not that suite needed to be updated. Based on previous PR feedback (which indicated that my unit test suite updates seemed sufficient), I have removed that commit (and the revert which followed it) from this PR.

> 2. What's up with the Rubocop rule changes? Why were _any_ Rubocop rule changes required?

TL;DR: These were changes related to #653 but not related directly to #644, so I have removed such unrelated changes from this PR.

Background:

These were mostly opportunistic "leave it as good or better than you found it" attempts to reduce the (large) quantity of offenses ignored/excluded by the `.rubocop_todo.yml` file (work which I subsequently created #653 to track). Both the changes to `.rubocop.yml` and the inline exclusions were my attempts to make the project's explicit rubocop rules match the project's existing design and implementation, in cases where the existing code was in direct conflict with rubocop default configurations. I.e. I was trying to make the rubocop config more accurately reflect certain non-rubocop-default project implementation choices, rather than leaving them as "ignored offenses to be fixed later".

However, for the sake of PR clarity and based on your feedback, I have removed all of those changes from this PR.

> 3. Why does this patch series contain a change to the version of rspec? Was that a _required_ part of adding this feature? If it wasn't, then that change should not be part of this patch series.

TL;DR: I have removed that code from this PR.

Background:

I encountered a spec failure which was difficult to understand due to rspec truncating the output. This is a known issue with rspec, and there is a semi-official "known fix" for it which requires a newer version of rspec than the one previously configured for this project.

However, after upgrading rspec, the "known fix" didn't work. So technically, the rspec upgrade wasn't required as part of this PR, since I didn't actually enable the not-yet-working "known fix". 

I had originally left the rspec upgrade in this PR since the effort to upgrade had already been done, and I had assumed that a newer rspec version would be preferred. However, I have now removed that upgrade from this PR, per your feedback.

> 4. Some of this patch series has parts that _just_ add or remove a blank line. These changes should not be in this patch series.

These were directly related to the aforementioned opportunistic rubocop offense fixes, and have been removed from this PR.

> 5. What's up with the return type annotations that are added to _some_ functions? For an example, consider the additional annotations that begin here: https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/files#diff-6eaadd97bb87d9f3b51e3214a0aa347de08a67afae0923b61b84cad6073b50b4R1955 Much of the rest of the file doesn't appear to have those... why were they added to otherwise untouched code? If you wanted to add type annotations to all of the functions in the CPI, that would be _lovely_... but that sort of change should be in a separate PR, not in this one.

TL;DR: I have removed that code from this PR.

Background:

There were a few utility methods called by code related to this PR which had no doc comments. And some of those methods' implementations had changed over time so that their method names had become inaccurate. I added the return type annotations to a few such methods while I was trying to understand the behavior of the original code (before I modified any of the behavior).

Since there were also many other un-annotated methods within `azure_client.rb` which also had names which were non-indicative of the return type (thus requiring a maintainer to read each method's implementations to understand what type of value would be returned), I opportunistically added `@return` annotations to many of those methods, assuming that such annotations would be better than completely un-annotated methods. However, per your feedback, I have removed from this PR the opportunistically-added annotations for the methods which were otherwise unmodified and unrelated to this PR.

> 6. Does this entirely-commented-out block actually do anything? https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/files#diff-1a1fbe20391f7dce08027421a87d3efb885328926612e3823ccf57a05259b7f0R172 If it does, I _really_ wanna know... Ruby is magical sometimes, and I'm always excited/horrified to learn of yet another way it can appear to do one thing, but actually do another. If it doesn't actually do something, you should remove it and also do a careful sweep to ensure that you haven't put _other_ commented-out dead code into this patch series.

TL;DR: I have removed that code from this PR.

Background:

That was the code for the aforementioned (see comment above RE: point # 3) "known fix" for the rspec truncation issue, which turned out to not work. Since the known issue is likely to be a problem again in future, I had left the disabled code (plus explanatory comments) for the not-yet-working fix, in case I or a future maintainer needed to attempt to et the fix working in future.

> 7. Looking at the individual commits, I would have an _extremely_ difficult time reviewing this PR. In fact, I had to have GitHub squash them down into a single diff so that I could understand the changes. That is a _very strong sign_ that anyone twelve months in the future who's trying to make sense of this part of the Git history will have a _very bad time_.
> 
> An ideal commit should contain changes that are logically grouped together, with an explanation of what was done and why.

I have attempted to explain both the what and the why in the messages of all 6 of the remaining commits (referencing #644 where appropriate, but without completely re-explaining the context captured in the description and comments of that issue).

Hopefully the new messages are satisfactory. I looked at many past commits within this repo for examples, but nearly all of them had very minimal, terse commit messages. So I mostly used your commit message guidance above.

> This helps future folks who don't have any context on what was changed in the code and why (which -BTW- this future person _very often_ ends up being _you_) understand the what and why of the code.
> 
> Please do squash this PR down to just a few commits that someone who has nearly-zero context on _what_ was done in this PR and _why_ it was done can read and understand. Remove the unnecessary whitespace changes, either help me to understand why you made RuboCop rules changes, or just undo those changes, and address the other stuff I brought up.

I believe each of the issues you previously raised have all now been fully addressed/resolved. If any were not, please let me know.

> _For now, I strongly NAK merging this PR as-is._
> 
> As an aside: if you're like me, and you make a ton of "checkpoint" commits while working on something big and/or unfamiliar, do feel free to keep those in a local branch for as long as you like. I -personally- find that keeping those commits around for quite a while _does_ help _me_ to find and fix errors in the code that I wrote... but I also recognize that the very many commits are only helpful to me _because_ I remember much of what I was thinking about when I authored those commits. I _also_ recognize that that mess of commits are _nearly completely_ useless to folks who lack that context. Other folks need the changes bundled up in larger chunks, with notes attached in the commit messages that explain all of the context that I had in my head while I was making those changes.

Thank you (again) for the time you took in providing detailed PR feedback. It was helpful and much appreciated.

Justin-W commented:
Thanks also to @bgandon for his additional tips and advice.

rkoster commented:
@klakin-pivotal are you satisfied with the changes made?

klakin-pivotal commented:
I hope to get some time this afternoon to review the changes. If I do not get that time, then I will make the time on Friday.

klakin-pivotal commented:
Overall, this PR looks good. Thanks for taking the time to reorganize it and clean it up.

If the folks who did the technical-functionality review are still fine with that aspect of the PR, and the few things I mention below get fixed, please feel free to merge it, whoever is in charge of merging it. (And if that "whoever" is me, let me know and I'll push the button on Tuesday (as Monday is a holiday for me).)

The remaining issues:

## Commented-out sections of code:

These commented-out sections of code should be removed, if they don't serve as documentation. If they _do_, then do accept my apologies for not noticing their significance:

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/c099e1f289474f43a27154ff81d35d331979aa7d#diff-5d4d5c53bf030aa246140de1d33e1f300b0093c44e87df37e1f5bd73f2c9508eR151

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/4ced629e8b70bab1bc6eb32223824420a79ac33c#diff-9850a43e9c57d5d53acbfe53ce7dfd26ff1ecbc939f05079faf1f76873e31780R459-R462

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/4ced629e8b70bab1bc6eb32223824420a79ac33c#diff-9850a43e9c57d5d53acbfe53ce7dfd26ff1ecbc939f05079faf1f76873e31780R911-R914

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/4ced629e8b70bab1bc6eb32223824420a79ac33c#diff-9850a43e9c57d5d53acbfe53ce7dfd26ff1ecbc939f05079faf1f76873e31780R603-R606

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/4ced629e8b70bab1bc6eb32223824420a79ac33c#diff-9850a43e9c57d5d53acbfe53ce7dfd26ff1ecbc939f05079faf1f76873e31780R620-R623

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/4ced629e8b70bab1bc6eb32223824420a79ac33c#diff-9850a43e9c57d5d53acbfe53ce7dfd26ff1ecbc939f05079faf1f76873e31780R889-R892

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/4ced629e8b70bab1bc6eb32223824420a79ac33c#diff-9850a43e9c57d5d53acbfe53ce7dfd26ff1ecbc939f05079faf1f76873e31780R1036-R1039

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/4ced629e8b70bab1bc6eb32223824420a79ac33c#diff-9850a43e9c57d5d53acbfe53ce7dfd26ff1ecbc939f05079faf1f76873e31780R1048-R1051

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/fba423041845f11c6bdeb8c50d7b0151b16fa0cb#diff-0f27965b66475a2daee83501a79ec969ed15ce1124fd6bac0bc899e857485b9aR231

https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/c099e1f289474f43a27154ff81d35d331979aa7d#diff-0f27965b66475a2daee83501a79ec969ed15ce1124fd6bac0bc899e857485b9aR371

All of the places that `# 'resource_group_name' => 'fake-rg-name',`, `# 'backend_pool_name' => 'fake-pool2-name'`, or `# 'backend_pool_name' => 'fake-ag2-pool-2-name'` appear in `dynamic_public_ip_spec.rb`: <https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651/commits/4ced629e8b70bab1bc6eb32223824420a79ac33c#diff-f32b10482e6d4d856d908fa2c875e88c5ade80f3d60c35c754a966f2676f8f2f>

Please don't feel obligated to re-work the commits that contain the commented-out sections of code. Adding another commit to the PR that removes any that should be removed seems fine to me.

## Comments added that have TODO in them

I'd ask that other folks who have previously evaluated the technical side of PR evaluate whether the TODO comments in the PR are comments that indicate that there are additional tests that should be added _now_, are comments that are useful to future test-writers, or are comments that should be removed.

Justin-W commented:
> These commented-out sections of code should be removed, if they don't serve as documentation. If they _do_, then do accept my apologies for not noticing their significance:

These did seem to have documentation value to me, but since that wasn't apparent to others I've removed all of the referenced commented sections.

> I'd ask that other folks who have previously evaluated the technical side of PR evaluate whether the TODO comments in the PR are comments that indicate that there are additional tests that should be added _now_, are comments that are useful to future test-writers, or are comments that should be removed.

The only TODO comments that I added were the `# TODO: Add more expectations here? The expects above only verify that the VM was created, ...` comments in `dynamic_public_ip_spec.rb`. The comment added on line 550 relates to a unit test that predates this PR's changes. All of the other TODO comments are in new unit tests which are derived from that original unit test (basically the same test, but using different configs). I wasn't sure how to fix the expectations in the original test, and therefore the new tests which are derived from it also use the same type of expectations as the original.

IMO, leaving these TODOs as suggestions for future work does no harm, since it does not worsen the state of the existing functionality or tests (the TODOs merely calls out an opportunity for future improvement).

Resolving the TODOs immediately (by improving the expectations of the marked tests) would obviously be better, but I'd need some guidance and/or an example of how to implement more precise expectations for the tests in that file, since these tests relate to a portion of the CPI's functionality about which I have limited understanding and (as I mentioned above) I'm not sure how to correctly resolve these TODOs without additional guidance.

klakin-pivotal commented:
@beyhan @ramonskie If you guys already did a technical-correctness review on the PR, would you mind re-checking this PR? Many commits have been merged together, and some of their contents have been changed, so it'd be good if you could ensure that the functionality remains correct.

Check this comment for a summary of what I thought would be good to do: <https://github.com/cloudfoundry/bosh-azure-cpi-release/pull/651#issuecomment-1013561004>

(I'm tagging both of you, because it's not clear which of you did a technical review... so if this doesn't apply to one of you, then feel free to ignore this comment.)

beyhan commented:
I will be able to go over again most probably tomorrow.

RobertBoyd commented:
> I will be able to go over again most probably tomorrow.

Thank you both @beyhan  @klakin-pivotal -  much appreciated. 

Given the amount of time this PR has been open with holidays and such, we've now moved onto another thread of work. Hopefully we can tie a bow on this sometime soon.

Justin-W commented:
@beyhan Can we get an update on this please?

beyhan commented:
I re-checked the pr and the logic to support multiple AGWs/LBs still looks good. @bgandon you requested changes which should be implemented now. Could you please re-check?

Justin-W commented:
@klakin-pivotal @rkoster I think this can be merged now, correct? As far as I know, all previously-raised concerns have now been resolved.

rkoster commented:
Yes you are right. Thanks @Justin-W for all the work you have put into this PR!
